### PR TITLE
Change middleware Use to type-check and compile ahead of time

### PR DESCRIPTION
**Pull request by Shane32**
Simple performance enhancement to existing code.

### CODE REVIEW

**Comment by Shane32** [comment]
I would suggest replacing this code with an IFieldMiddleware interface rather than dynamic binding.  See #1443

**Comment by sungam3r** [comment]
Just for information - do you know that this code is usually executed only once when the `Schema` is initialized?

**Comment by Shane32** [comment]
I figured as much, but wasn’t sure. I did think that this way, any compilation error would be raised when Use was called rather than later on — e.g. when the schema is initialized.

**Comment by sungam3r** [comment]
> I did think that this way, any compilation error would be raised when Use was called rather than later on — e.g. when the schema is initialized.

In principle, this is better.

File: src/GraphQL/Instrumentation/FieldResolverBuilderExtensions.cs [hunk 1]
```diff
@@ -10,39 +11,50 @@ public static class FieldResolverBuilderExtensions
     {
         private const string InvokeMethodName = "Resolve";
 
-        public static IFieldMiddlewareBuilder Use<T>(this IFieldMiddlewareBuilder builder) => Use(builder, typeof(T));
+        public static IFieldMiddlewareBuilder Use<T>(this IFieldMiddlewareBuilder builder) where T : new() => Use(builder, typeof(T));
```
Comment (sungam3r) [inline src/GraphQL/Instrumentation/FieldResolverBuilderExtensions.cs]: > where T : new()

nice catch

**Comment by sungam3r** [comment]
@Shane32 We also can think about what to do if middleware implements `IDisposable`. Since we ourselves create middleware, we must also destroy it, checking for `IDisposable` support, but this is certainly not the concern of this PR.

**Comment by sungam3r** [comment]
> Simple performance enhancement to existing code.

And strictly speaking, these are not performance enhancements. If you specifically conduct a test, I’m sure that it will show a slowdown, since the method is actually called once, and the cost of dynamic compilation is high. Dynamic compilation gives a performance gain only when the additional cost of it is negligible compared to the cost of repeatedly calling the method through reflection.

But still it seems to me that this is a step in the right direction due to the argument already mentioned above in favor of early detection of errors.

**Comment by Shane32** [comment]
I just realized something: Assuming there the builder code only ran once as you say (during schema initialization), then there is a significant difference between the new and old code -- the new code creates a new instance of the type every time the middleware executes, whereas the old code creates an instance once and calls the Resolve function repeatedly every time the middleware executes.

I didn't realize this when I wrote the code.  Should the code be modified to create the instance once, when the schema is initialized?

**Comment by sungam3r** [comment]
Wow. You messed up something. Check the code again. Both options create a middleware once, just the second one activates it delayed, at the execution stage, and not the installation stage. As you said before.

**Comment by Shane32** [comment]
The compiled expression is equivalent to:
``` C#
//func = (context, next) => (new <middleware>()).Resolve(context, next);
```
So within this:
``` C#
            return builder.Use(next =>
            {
                return context => func(context, next);
            });
```
We can see that every time the lambda is executed, an instance of the middleware will be created.

The behavior, if desired to be fixed, is fixed in #1463

**Comment by sungam3r** [comment]
I'm surprised that I could be so wrong

**Pull request merged**